File: src/tanotly/clipboard.py
"""Clipboard utilities for Tanotly."""

import re
import subprocess
import tempfile

from .data.models import DataNode


def copy_to_clipboard(content: str) -> tuple[bool, str]:
    """Copy text to clipboard. Returns (success, message)."""
    commands = [
        ['xclip', '-selection', 'clipboard'],  # Linux
        ['pbcopy'],                              # macOS
        ['clip'],                                # Windows
    ]

    for cmd in commands:
        try:
            subprocess.run(cmd, input=content.encode(), check=True, capture_output=True)
            return True, "Copied to clipboard"
        except (FileNotFoundError, subprocess.CalledProcessError):
            continue

    # Fallback: save to temp file
    with tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.txt') as f:
        f.write(content)
        return False, f"Saved to {f.name}"


def format_tree_text(tree_node, prefix: str = "", is_last: bool = True) -> str:
    """Recursively format tree nodes as plain text."""
    result = ""
    if tree_node.data:
        connector = "└── " if is_last else "├── "
        label = tree_node.label
        label_text = label.plain if hasattr(label, 'plain') else re.sub(r'\[.*?\]', '', str(label))
        result += prefix + connector + label_text + "\n"

        if tree_node.children:
            extension = "    " if is_last else "│   "
            for i, child in enumerate(tree_node.children):
                result += format_tree_text(child, prefix + extension, i == len(tree_node.children) - 1)
    else:
        for i, child in enumerate(tree_node.children):
            result += format_tree_text(child, "", i == len(tree_node.children) - 1)
    return result


def format_node_info(node: DataNode) -> str:
    """Format node information as plain text."""
    lines = [
        f"Node: {node.name}",
        f"Type: {node.node_type.value}",
        f"Path: {node.path}",
    ]

    if node.metadata:
        lines.append("\nMetadata:")
        for key, val in node.metadata.items():
            lines.append(f"  {key}: {val}")

    if node.attributes:
        lines.append("\nAttributes:")
        for key, val in node.attributes.items():
            lines.append(f"  {key}: {val}")

    return "\n".join(lines) + "\n"

File: src/util/clipboard.rs
//! Clipboard integration.

use crate::data::DataNode;
use crate::error::Result;
use arboard::Clipboard;

/// Copy text to clipboard, with fallbacks for WSL and headless Linux.
fn copy_to_clipboard(text: &str) -> Result<()> {
    // 1. arboard (works when X11/Wayland display is available)
    if let Ok(mut cb) = Clipboard::new() {
        if cb.set_text(text).is_ok() {
            return Ok(());
        }
    }

    // 2. WSL: clip.exe (Windows clipboard, always available in WSL2)
    if try_pipe_to_cmd("clip.exe", text) {
        return Ok(());
    }

    // 3. Wayland (wl-copy)
    if try_pipe_to_cmd("wl-copy", text) {
        return Ok(());
    }

    // 4. X11 (xclip)
    if try_pipe_to_cmd_args("xclip", &["-selection", "clipboard"], text) {
        return Ok(());
    }

    Err(crate::error::CoriolisError::Clipboard(
        arboard::Error::Unknown {
            description: "No clipboard backend available. On WSL ensure clip.exe is accessible or set DISPLAY.".to_string(),
        },
    ))
}

fn try_pipe_to_cmd(cmd: &str, text: &str) -> bool {
    try_pipe_to_cmd_args(cmd, &[], text)
}

fn try_pipe_to_cmd_args(cmd: &str, args: &[&str], text: &str) -> bool {
    use std::io::Write;
    use std::process::{Command, Stdio};

    let mut child = match Command::new(cmd).args(args).stdin(Stdio::piped()).spawn() {
        Ok(c) => c,
        Err(_) => return false,
    };

    if let Some(stdin) = child.stdin.take() {
        let mut writer = std::io::BufWriter::new(stdin);
        let _ = writer.write_all(text.as_bytes());
    }

    child.wait().map(|s| s.success()).unwrap_or(false)
}

/// Copy tree structure to clipboard.
pub fn copy_tree_structure(node: &DataNode, file_name: Option<&str>) -> Result<()> {
    let mut text = String::new();

    if let Some(name) = file_name {
        text.push_str(&format!("Tree Structure: {}\n", name));
    } else {
        text.push_str("Tree Structure\n");
    }

    text.push_str(&"=".repeat(80));
    text.push_str("\n\n");

    text.push_str(&format_tree_recursive(node, "", true));

    copy_to_clipboard(&text)
}

/// Copy node information to clipboard.
pub fn copy_node_info(node: &DataNode) -> Result<()> {
    let mut text = format!("Node: {}\n", node.name);
    text.push_str(&format!("Path: {}\n", node.path));
    text.push_str(&format!("Type: {:?}\n", node.node_type));

    if let Some(ref shape) = node.shape {
        text.push_str(&format!("Shape: {:?}\n", shape));
    }

    if let Some(ref dtype) = node.dtype {
        text.push_str(&format!("DType: {}\n", dtype));
    }

    if !node.attributes.is_empty() {
        text.push_str("\nAttributes:\n");
        for (key, value) in &node.attributes {
            text.push_str(&format!("  {}: {}\n", key, value));
        }
    }

    if !node.metadata.is_empty() {
        text.push_str("\nMetadata:\n");
        for (key, value) in &node.metadata {
            text.push_str(&format!("  {}: {}\n", key, value));
        }
    }

    copy_to_clipboard(&text)
}

fn format_tree_recursive(node: &DataNode, prefix: &str, is_last: bool) -> String {
    let mut result = String::new();

    let connector = if is_last { "└── " } else { "├── " };
    result.push_str(&format!("{}{}{}\n", prefix, connector, node.display_name()));

    let new_prefix = format!("{}{}   ", prefix, if is_last { " " } else { "│" });

    for (i, child) in node.children.iter().enumerate() {
        let is_last_child = i == node.children.len() - 1;
        result.push_str(&format_tree_recursive(child, &new_prefix, is_last_child));
    }

    result
}
